### ISSUE TITLE
Excon always prefers ipv4 addresses over ipv6

### ISSUE THREAD

**Issue by davidtaylorhq** [excon/excon#794]
Excon leans on `Resolv.each_address`:

https://github.com/excon/excon/blob/226924e59db71a1c8ce54a682cb343c55be44219/lib/excon/socket.rb#L109

This will [always yield ipv4 addresses before ipv6 addresses](https://github.com/ruby/resolv/blob/4a6c1ec1a6284218fe548ea13dc341221dbc227e/lib/resolv.rb#L402-L405). That means that Excon will always prefer ipv4 over ipv6.

`Net::HTTP` uses [`Socket.tcp`](https://github.com/ruby/ruby/blob/aa490f9442c32cd0e1e449ac817f410bd5924c8b/ext/socket/lib/socket.rb#L632), which in turn uses `Addrinfo.getaddrinfo`. That returns addresses in an order defined by the OS (e.g. as specified in `/etc/gai.conf`).

It would be great if Excon could respect the system configuration for ipv4/ipv6 precedence.

**Comment by SamSaffron** [comment]
The rabbit hole is very very deep. There are mountains of rules here: 

https://www.ietf.org/rfc/rfc3484.txt

You are meant to apply them on the list you get back from DNS depending on the source address. (and then Linux also allow for the gai.conf for overriding - which we could support) 

To be honest just adding an optional shuffle via config to Excon at this point in time would probably be the the simplest workaround. Implementing the RFC with overrides is the gold class, but we are talking weeks of development to build that ... probably...

**Comment by geemus** [comment]
@SamSaffron Thanks, that does seem overwhelming. So you think having something simple and direct might be better, ie allowing you to set an option that specifies whether you prefer v4 or v6? I could definitely see how that would be simpler and avoid most of the concerns you mentioned.

@davidtaylorhq - would that work for you? Or is following `/etc/gai.conf` a requirement as well? I'd like to support options, but trying also to keep this from getting too complicated if we can avoid it.

**Comment by SamSaffron** [comment]
> ie allowing you to set an option that specifies whether you prefer v4 or v6?  

I think the simplest thing is probably to just allow people to plug in a resolver, maybe revert to the old implementation by default and provide a few options: 

```
Excon.resolver = lambda { |host| MyDNSThing.resolv(host) } 
```

This would open the door to a bunch of things:

- People that want a resolver based on `Resolv` can use that 
- People that want a resolver based on `Socket.gethostbyname` can use that 
- People that want to mitigate against DoS can use `Socket.gethostbyname`  with a timeout implementation 

and so on 

At Discourse we are opting for `Socket.gethostbyname` with a custom monitor thread that times out long requests. This is probably the safest default, but it is complicated code.

**Comment by geemus** [comment]
@SamSaffron Ah, thanks, yeah I think something like that makes sense. I'll try to find some time to figure that out in the not too distant future.

**Comment by geemus** [comment]
I believe this should now be resolved thanks to @bruno-b-martins work in #834.
